cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/GoalOptimizer.java

```diff
@@ -117,6 +117,52 @@ public GoalOptimizer(KafkaCruiseControlConfig config,
     _proposalComputationTimer = dropwizardMetricRegistry.timer(MetricRegistry.name("GoalOptimizer", "proposal-computation-timer"));
   }
 
+  /**
+   * Generate unique goal priority for each of proposal precomputing thread.
+   */
+  private void populateGoalByPriorityForPrecomputing() {
+    List<List<Goal>> shuffledGoalList = new ArrayList<>();
+    // Guarantee one thread is working on the original goal priority.
+    shuffledGoalList.add(new ArrayList<>(_goalsByPriority.values()));
+
+    Integer goalCount = _goalsByPriority.values().size();
+    Integer goalIndex = goalCount - 2;
+    Integer numShuffledGoalToGenerate = numProposalComputingThreads() - 1;
+    while (numShuffledGoalToGenerate > 0) {
+      List<List<Goal>> tmpShuffledGoalList = new ArrayList<>();
+      for (int i = 0; i < shuffledGoalList.size() && numShuffledGoalToGenerate > 0; i++) {
```

Comment:
The algorithm needs more comments to clarify the logic.

Reply:
added comment.